Firmware/ultralcd.cpp

```diff
@@ -8959,6 +8967,43 @@ void lcd_experimental_menu()
     MENU_END();
 }
 
+void reprint_from_eeprom() {
+	char cmd[30];
+	char filename[13];
+	uint8_t depth = 0;
+	char dir_name[9];
+
+	isPrintFinished=false;
+
+	//cmdqueue_reset();
+
+	depth = eeprom_read_byte((uint8_t*)EEPROM_DIR_DEPTH);
+	
+	MYSERIAL.println(int(depth));
+	for (int i = 0; i < depth; i++) {
+		for (int j = 0; j < 8; j++) {
+			dir_name[j] = eeprom_read_byte((uint8_t*)EEPROM_DIRS + j + 8 * i);
+		}
+		dir_name[8] = '\0';
+		MYSERIAL.println(dir_name);
+		// strcpy(dir_names[i], dir_name);
+		card.chdir(dir_name, false);
+	}
+
+	for (int i = 0; i < 8; i++) {
+		filename[i] = eeprom_read_byte((uint8_t*)EEPROM_FILENAME + i);
+	}
+	filename[8] = '\0';
+
+	MYSERIAL.print(filename);
+	strcat_P(filename, PSTR(".gco"));
```

Comment:
- are you sure the printed file always ends with a ".gco"?
- how about "1.g"?

Reply:
* I use the same logic that the one developed on the function **restore_print_from_eeprom()**
* As far as I know the eeprom value **EEPROM_FILENAME** only can contain 8 chars, and these chars are stored from the function **menu_action_sdfile()** on Marlin_main.cpp, where it takes the filename and assume always that the extension is _".gco"_ 

> ```
> static void menu_action_sdfile(const char* filename)
> {
>   loading_flag = false;
>   char cmd[30];
>   char* c;
>   bool result = true;
>   sprintf_P(cmd, PSTR("M23 %s"), filename);
>   for (c = &cmd[4]; *c; c++)
>     *c = tolower(*c);
> 
>   const char end[5] = ".gco";
> 
>   //we are storing just first 8 characters of 8.3 filename assuming that extension is always ".gco"
>   for (uint_least8_t i = 0; i < 8; i++) {
> 	  if (strcmp((cmd + i + 4), end) == 0) {
> 		  //filename is shorter then 8.3, store '\0' character on position where ".gco" string was found to terminate stored string properly
>  		  eeprom_write_byte((uint8_t*)EEPROM_FILENAME + i, '\0');
> 		  break;
> 	  }
> 	  else {
> 		  eeprom_write_byte((uint8_t*)EEPROM_FILENAME + i, cmd[i + 4]);
> 	  }
>   }
> ```

* There is a similar situation when the printer tries to recover from a power disruption because the function **restore_print_from_eeprom()** always assume that the extension will be _".gco"_  and concatenate the extension _".gco"_ with the filename 
* How about 1.g? The reprint will not work and I suppose that the recovery from a power outage will not work also because will not be able to find the file.

Reply:
Oh, good point, now I got why you made it this way.
Honestly, as the new Reprint function will be immediately available to users (read much more used than Power Panic), it should really work with all of the supported file extensions ... that would also probably fix the issue in Power Panic as well.
I'm not sure how to fix restore_print_from_eeprom() correctly yet.

Reply:
Maybe a solution could be to check if the file exists on the SD with the _".gco"_ and if the file is not found try to locate it without the extension. This will work if the filename is less than 8 chars but if the name is longer I'm not sure about the behaviour, is something that should be tested. What do you think?

Reply:
@DRracer I was thinking about this issue and maybe a solution could be to use a compression library in order to compress the string of the whole path (including the filename) before to store on the eeprom and decompress the string on a reprint situation or a recovery from a power outage.  The bad side is that needs compute efforts and free space for the library, also must be sure that the compressed string does not exceed the size reserved on the eeprom to store it. 
There are some libraries like https://github.com/siara-cc/Unishox_Arduino_Progmem_lib which could be studied.

Reply:
I think we can fix all these cases (reprint, restore and menu select from sd) by adding an additional function that starts the print via M24 starting from a file prefix. How we fix the extension (either by listing or by trying a second time with .g) would be trivial then

Reply:
By now I have implemented the following solution: 

```
	strcpy(altfilename,filename);
	if (!card.FileExists(altfilename))
	{
		strcat_P(filename, PSTR(".gco"));
		if (card.FileExists(filename))
		{
			strcpy(altfilename,filename);
		}else
		{
			strcat_P(altfilename, PSTR(".g"));
		}
	}
	MYSERIAL.print(altfilename);
	sprintf_P(cmd, PSTR("M23 %s"), altfilename);
```

I know that is not the more elegant solution but will work with .g and .gcode extensions. In order to know if the file exists or not I have implemented the _card.FileExists()_ function.